File: LeetCode/2183. Count Array Pairs Divisible by K/solution.py
    def coutPairs(self, A, k):
        cnt = Counter(math.gcd(a, k) for a in A)
        res = 0
        for a in cnt:
            for b in cnt:
                if a <= b and a * b % k == 0:
                    res += cnt[a] * cnt[b] if a < b else cnt[a] * (cnt[a] - 1) // 2
        return res


class Solution:
    def coutPairs(self, nums: List[int], k: int) -> int:
        nums.sort()
        def gcd(a, b):
            if b == 0:
                return a 
            if a < b:
                return gcd(b, a)
            return gcd(b, a%b)

        cnt = 0
        cc = Counter()
        for i, x in enumerate(nums):
            g = gcd(x, k)
            g = k//g 
            cnt += cc[g]
            q = 1
            # print(i,x,cnt,cc)
            while q*q <= k:
                if k%q == 0:
                    if x%q == 0:
                        cc[q] += 1
                    p = k//q
                    if x%p == 0 and p != q:
                        cc[p] += 1
                q += 1
        return cnt 


class Solution:
    def coutPairs(self, nums: List[int], k: int) -> int:
        N, output = len(nums), 0
        divisors = []
        counter = Counter()
        
        for i in range(1, k + 1):
            if k % i == 0:
                divisors.append(i)
        
        for i in range(0, N):
            remainder = k // math.gcd(k, nums[i])
            output += counter[remainder]
            for divisor in divisors:
                if nums[i] % divisor == 0:
                    counter[divisor] += 1
            
        return output


class Solution:
    def coutPairs(self, nums: List[int], k: int) -> int:
        def gcd(a, b):
            if b == 0:
                return a 
            return gcd(b, a%b)

        cnt = 0
        cc = Counter(nums)
        mx = max(nums)
        mx = max(mx, k)
        for f in range(1,mx+1):
            if k%f:
                continue
            for fs in range(f*2,mx+1,f):
                cc[f] += cc[fs]
        for x in nums:
            g = gcd(x, k)
            g = k//g 
            cnt += cc[g]
        for x in nums:
            if x*x%k == 0:
                cnt -= 1
        return cnt//2


class Solution:
    def coutPairs(self, nums: List[int], k: int) -> int:
        def gcd(a, b):
            if b == 0:
                return a 
            return gcd(b, a%b)

        cnt = 0
        cc = Counter(nums)
        cc2 = Counter()
        mx = max(nums)
        for f in range(1,k+1):
            if f*f > k:
               break
            if k%f:
                continue
            for fs in range(f,mx+1,f):
                cc2[f] += cc[fs]
            f2 = k//f 
            if f2 == f:
                continue
            for fs in range(f2,mx+1,f2):
                cc2[f2] += cc[fs]
        # print(cc2)
        for x in nums:
            g = gcd(x, k)
            g = k//g 
            cnt += cc2[g]
            if (x*x)%k == 0:
                cnt -= 1
        return cnt//2

File: LeetCode/2183. Count Array Pairs Divisible by K/solution.cpp
class Solution {
public:
    long long coutPairs(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        for (int& nu: nums) {
            mp[gcd(nu, k)]++;
        }
        long long res = 0;
        for (auto& [a, c1]: mp) {
            for (auto & [b, c2]: mp) {
                if (a <= b && a*(long) b %k == 0) {
                    res += a < b?(long)c1*c2:(long)c1*(c1-1)/2;
                }
            }
        }
        return res;    
    }
};

class Solution {
public:
    long long coutPairs(vector<int>& nums, int k) {
        int mx = k;
        for (int x : nums) mx = max(mx, x);

        // 统计每个数的倍数出现的次数
        vector<int> cnt(mx + 1), cnt2(mx + 1);
        for (int x : nums) cnt[x]++;
        // 为什么这个算法是 O(nlnn) 的？因为这个算法的循环次数是 n(1 + 1/2 + 1/3 + ...)，由调和级数可知括号内趋向 lnn
        for (int i = 1; i*i <= k; i++){
            if (k%i) continue;
            for (int j = i; j <= mx; j += i) cnt2[i] += cnt[j];
            int ii = k/i;
            if (ii == i) continue;
            for (int j = ii; j <= mx; j += ii) cnt2[ii] += cnt[j];
        }

        long long ans = 0;
        // 对于每个数统计与它形成好二元组的数有几个
        for (int x : nums) ans += cnt2[k / __gcd(k, x)];
        // 排除自己和自己形成好二元组的情况
        for (int x : nums) if (1LL * x * x % k == 0) ans--;
        return ans / 2;
    }
};


// 作者：tsreaper
// 链接：https://leetcode-cn.com/problems/count-array-pairs-divisible-by-k/solution/shu-xue-ji-shu-by-tsreaper-7ok1/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。




class Solution {
public:
    long long coutPairs(vector<int>& nums, int k) {
        long long ans = 0;
        int mx = k;
        for (int x : nums) mx = max(mx, x);

        vector<int> cnt(mx + 1);
        for (int x : nums) {
            int g = k/__gcd(x, k);
            ans += cnt[g];
            for (int q = 1; q*q <= k; q++){
                if (k%q == 0){
                    if (x%q == 0) cnt[q]++;
                    int p = k/q;
                    if (p == q) continue;
                    if (x%p == 0) cnt[p]++;
                }
            }
        }
      
        return ans;
    }
};


const int mx = 100001;
vector<int> divisors[mx];

int init = []() { // 预处理每个数的所有因子，时间复杂度 O(MlogM)，M=1e5
    for (int i = 1; i < mx; ++i)
        for (int j = i; j < mx; j += i)
            divisors[j].push_back(i);
    return 0;
}();

class Solution {
public:
    long long coutPairs(vector<int> &nums, int k) {
        long long ans = 0;
        unordered_map<int, int> cnt;
        for (int v : nums) {
            ans += cnt[k / gcd(v, k)];
            for (int d : divisors[v])
                ++cnt[d];
        }
        return ans;
    }
};


作者：endlesscheng
链接：https://leetcode-cn.com/problems/count-array-pairs-divisible-by-k/solution/tong-ji-yin-zi-chu-xian-ci-shu-by-endles-t5k8/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。


class Solution {
public:
    long long coutPairs(vector<int> &nums, int k) {
        vector<int> divisors; 
        for (int d = 1; d * d <= k; ++d) { // 预处理 k 的所有因子
            if (k % d == 0) {
                divisors.push_back(d);
                if (d * d < k) divisors.push_back(k / d);
            }
        }
        long long ans = 0;
        unordered_map<int, int> cnt;
        for (int v : nums) {
            ans += cnt[k / gcd(v, k)];
            for (int d : divisors)
                if (v % d == 0) ++cnt[d];
        }
        return ans;
    }
};


作者：endlesscheng
链接：https://leetcode-cn.com/problems/count-array-pairs-divisible-by-k/solution/tong-ji-yin-zi-chu-xian-ci-shu-by-endles-t5k8/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。